### PR TITLE
cmd/torusblk: implemenet NBD TCP server

### PR DESCRIPTION
**Pull request by mischief**

### CODE REVIEW

File: cmd/torusblk/nbd.go [hunk 1]
```diff
@@ -104,3 +120,66 @@ func connectNBD(srv *torus.Server, f *block.BlockFile, target string, closer cha
...
+	srv *torus.Server
+}
+
+func (f *finder) FindDevice(name string) (nbd.Device, error) {
+	blockvol, err := block.OpenBlockVolume(f.srv, name)
+	if err != nil {
+		return nil, err
+	}
+
+	dev, err := blockvol.OpenBlockFile()
```
Comment (xiang90) [inline cmd/torusblk/nbd.go]: return blockvol.OpenBlockFile() ?

**Comment by mischief** [comment]
ptal

**Comment by mischief** [comment]
test flaked, see #212/#221

File: internal/nbd/nbdserver.go [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+
+	"github.com/coreos/pkg/capnslog"
+)
+
+var clog = capnslog.NewPackageLogger("github.com/coreos/torus", "nbd")
+
+const (
+	nbdMagic       uint64 = 0x4e42444d41474943
+	nbdOpts        uint64 = 0x49484156454F5054
+	nbdRequestSize        = (4 + 4 + 8 + 8 + 4)
```
Comment (xiang90) [inline internal/nbd/nbdserver.go]: a simple comment for the bytes like 4 4 8 8 4?

File: internal/nbd/nbdserver.go [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+	nbdOptExportName = 0x1
+	nbdOptAbort      = 0x2
+	nbdOptList       = 0x3
+
+	// option replies
+	nbdRepAck      = 0x1
+	nbdRepServer   = 0x2
+	nbdRepErrUnsup = 0x80000001
+
+	ioKeepAlive = 10 * time.Second
```
Comment (xiang90) [inline internal/nbd/nbdserver.go]: tcpKeepAlive?

File: internal/nbd/nbdserver.go [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+	c      *net.TCPConn
+	finder DeviceFinder
+	device Device
+	export string
+}
+
+func (c *NBDConn) errorf(format string, stuff ...interface{}) {
+	s := fmt.Sprintf(format, stuff...)
+	clog.Errorf("%s: %s: %s", c.c.RemoteAddr(), c.export, s)
+
```
Comment (xiang90) [inline internal/nbd/nbdserver.go]: remove extra empty line?

File: internal/nbd/nbdserver.go [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+}
+
+func (c *NBDConn) tracef(format string, stuff ...interface{}) {
+	s := fmt.Sprintf(format, stuff...)
+	clog.Tracef("%s: %s: %s", c.c.RemoteAddr(), c.export, s)
+
```
Comment (xiang90) [inline internal/nbd/nbdserver.go]: same here.

File: internal/nbd/nbdserver.go (hunk 2 above)
Comment (mischief) [inline internal/nbd/nbdserver.go]: unused now, ill drop it.

File: internal/nbd/nbdserver.go [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+	return s.l.Close()
+}
+
+type option struct {
+	opt  uint32
+	data []byte
+}
+
+type NBDConn struct {
+	c      *net.TCPConn
```
Comment (xiang90) [inline internal/nbd/nbdserver.go]: shall we change this to conn interface?

File: internal/nbd/nbdserver.go [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+}
+
+func (c *NBDConn) tracef(format string, stuff ...interface{}) {
+	s := fmt.Sprintf(format, stuff...)
+	clog.Tracef("%s: %s: %s", c.c.RemoteAddr(), c.export, s)
+
+}
+
+func (c *NBDConn) handler() error {
+	var dev Device
```
Comment (xiang90) [inline internal/nbd/nbdserver.go]:
```go
var (
    ....
)
```?

File: internal/nbd/nbdserver.go (hunk 6 above)
Comment (mischief) [inline internal/nbd/nbdserver.go]: can't keepalive a net.Conn, but i can drop that if you think it's better.

File: internal/nbd/nbdserver.go [hunk 8]
```diff
@@ -0,0 +1,315 @@
...
+				if err := c.optReply(nbdOptList, nbdRepServer, buf); err != nil {
+					return err
+				}
+			}
+
+			if err := c.optReply(nbdOptList, nbdRepAck, nil); err != nil {
+				return err
+			}
+		default:
+			if err := c.optReply(opt.opt, nbdRepErrUnsup, nil); err != nil {
```
Comment (xiang90) [inline internal/nbd/nbdserver.go]: trace unknown ops?

Comment (mischief) [inline internal/nbd/nbdserver.go]: errors are dumped in Serve()

File: internal/nbd/nbdserver.go [hunk 9]
```diff
@@ -0,0 +1,315 @@
...
+
+	if err := binary.Read(c.c, binary.BigEndian, &clientFlags); err != nil {
+		return err
+	}
+
+	c.tracef("reading options")
+
+	// read options
+optLoop:
+	for {
```
Comment (xiang90) [inline internal/nbd/nbdserver.go]: i feel we should probably make this for loop a separate func.

File: internal/nbd/nbdserver.go [hunk 10]
```diff
@@ -0,0 +1,315 @@
...
+}
+
+func writeFull(w io.Writer, buf []byte) error {
+	n, err := w.Write(buf)
+	if err != nil {
+		return err
+	}
+
+	if n != len(buf) {
+		return errors.New("short write")
```
Comment (xiang90) [inline internal/nbd/nbdserver.go]: does not write already do this?

Comment (xiang90) [inline internal/nbd/nbdserver.go]: or we should use io.ErrShortWrite. see https://golang.org/pkg/io/

File: internal/nbd/nbdserver.go [hunk 11]
```diff
@@ -0,0 +1,315 @@
...
+func (c *NBDConn) optReply(opt uint32, reptype uint32, data []byte) error {
+	rep := make([]byte, 8+4+4+4)
+
+	// option reply magic
+	binary.BigEndian.PutUint64(rep[0:8], uint64(0x3e889045565a9))
+	binary.BigEndian.PutUint32(rep[8:12], opt)
+	binary.BigEndian.PutUint32(rep[12:16], reptype)
+	binary.BigEndian.PutUint32(rep[16:20], uint32(len(data)))
+
+	if len(data) > 0 {
```
Comment (xiang90) [inline internal/nbd/nbdserver.go]: we do not need this checking i think. just always append?

File: internal/nbd/nbdserver.go [hunk 12]
```diff
@@ -0,0 +1,315 @@
...
+		rep = append(rep, data...)
+	}
+
+	return writeFull(c.c, rep)
+}
+
+func (c *NBDConn) getOpt() (*option, error) {
+	var optMagic uint64
```
Comment (xiang90) [inline internal/nbd/nbdserver.go]:
```go
var (
    ...
)
```

**Comment by xiang90** [comment]
minor issues. look good to me.

File: internal/nbd/nbdserver.go (hunk 6 above)
Comment (xiang90) [inline internal/nbd/nbdserver.go]: i know that. we already set keep alive before putting this into ndbconn. not a big deal anyway. it is up to you. probably we will only have tcp ndb conn in foreseeable future...

File: internal/nbd/nbdserver.go (hunk 8 above)
Comment (xiang90) [inline internal/nbd/nbdserver.go]: you are right.

**Comment by mischief** [comment]
ptal

**Comment by xiang90** [comment]
LGTM

**Pull request merged**